students/pinstripezebra/mailroomP3.py

```diff
@@ -0,0 +1,127 @@
+        
+            #if the user chooses to send a thank you to everyone
+            elif userAction == "3":
+                groupThanks(donationHistory)
+            #if none of the above options are chosen quits the program
+            elif userAction == 4:
+                print("Quiting program")
+                break
+        except ValueError:
+            print("The value entered must be 1, 2, 3, or 4")
+    
+              
+    
+#If the user selects to send a thank you letter to a single donor   
+def thankYou(nameRequest, donationAmount=0):
+    names = donationHistory.keys()    
+    
+    #Prompts the user to select a name if they enter "list"
+    while nameRequest == "list":
+        for i in range(len(list(names))):
+            print(list(names)[i])
+        nameRequest = input( "Please select a name: ")
+        
+    #If the user enters a name that exists 
+    if nameRequest in names:
+        try:
+            #Optional parameter donationAmount can be passed in for testing
+            #If it has not been passed in for testing asks user for it here
+            if donationAmount == 0:
+                donationAmount = int(input( "Please input a donation amount: "))
+                
+            donationHistory[nameRequest][0] = round(int(donationHistory[nameRequest][0]) + donationAmount, 2)
+            donationHistory[nameRequest][1] += 1
+            
+        #If the donation amount entered isnt valid
+        except TypeError:
+            print("Value entered must be an integer")
+            
+        except ValueError:
+            print("Value entered must be greater than 0")
+        
+    #If the user enters a name that doesnt exist
+    else:
+        try:
+            #If the code isnt being ran through testing code
+            if donationAmount == 0:
+                donationAmount = int(input("Please input a donation amount: "))
+                
+            donationHistory[nameRequest] = [donationAmount, 1]
+            
+        except TypeError:  
+            print("value entered must be an integer")
+        
+        except ValueError:
+            print("Value entered must be greater than 0")
+            
+    print('Dear {},\n\n Thank you for your generous donation of ${}, it will be put to good use. \n\nThe team'.format(nameRequest, donationAmount))
+    return donationHistory
+
+#If the user wants to print a report of the current donor information    
+def printReport(donationHistory):
+    #Have to expilicitly declare names as list, otherwise get 'dict_keys' object
+    names = list(donationHistory.keys())
+    firstColumn = len(max(names, key=len))
+    #The length of the string "Donor Name
+    secondColumn = 10
+    header = "Donor Name"+' '*max(firstColumn - 10,0)+ "|" +"Total Given" + "  |" + "Num Gifts" + "|" + "Average Gift"
+    print(header)
+    print("-"*len(header))
+    
+    #Loops over all the donors and prints one line for each
+    for i in range(len(names)):

```

Comment:
no need for `len()` here.  The `for` loop will already know to iterate on the length of `names`